---
title: "Release"
author: "Paul Deveau"
date: "`r Sys.Date()`"
output: html_vignette 
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Release}
  %\VignetteEncoding{UTF-8}
---
<h3<Release of X/9/2016 </h3>
<ul>
<li>Changes in weight computation: normalization by sum of weights; </li>
<li>Added <a href=https://www.jstatsoft.org/article/view/v040i06>DEoptim </a> solver for EM;</li>
<li>Flash QuantumClone handles polyploid regions;</li>
<li>Flash QuantumClone handles BIC, AIC or numeric correction of BIC for cluster selection.</li>

</ul>

<h3<Release of 9/9/2016 </h3>
<h4>MAY BE UNSTABLE</h4>
<ul>
<li> Added FlashQC : using hierarchical clustering and distance on the p-value of 2 variants being from the same distribution; </li>
<li> Updated parallel: initialization and number of clusters are run in parallel, not independently; </li>
<li> For rearranged genomes, after the first clustering, a second one is done after removing unlikely states; </li>
<li> Preclustering can now be done with FlashQC </li>
</ul>

<h3>Release of 31/08/2016</h3>
<ul>
<li>Rewrote this vignette in html </li>
<li>Added reproducible testing for PyClone, sciClone and k-medoid</li>
<li>Updated figure dpi for vignettes</li>
</ul>

<h3>Release of 19/08/2016:</h3>
<ul>
<li> Corrected NMI computation if a number of cluster is missing (for example cluster values can be 1:3, 5:7, cluster value 6 would be considered missing);</li>
<li> Changed the proof of concept pipeline to be easier to read and edit;</li>
<li> Changes in `Probability.to.belong.to.clone`: now returns filtered and unfiltered probabilities (clustering is on filtered probabilities).</li>
</ul>

<h3>Release of 16/08/2016:</h3>
<ul>
<li> Corrected hard clustering parameters; </li>
<li> Corrected `Probability.to.belong.to.clone`; </li>
<li> Gain from parallelization script. </li>
</ul>

<h3>Release of 12/08/2016:</h3>
<ul>
<li> Computed PoC for clustering quality; </li>
<li> Changes in `eval.fik` for an expected 5-10% gain in computational speed.</li>
</ul>
<h3>Release of 9/08/2016:</h3>
<ul>
<li> Removed 3D plots to prepare CRAN release </li>
</ul>
<h3>Release of 8/08/2016:</h3>
<ul>
<li> Removed rgl dependency (3D plots), causing QuantumClone to be unavailable on Ubuntu;</li>
<li> Updated pipeline tests;</li>
<li> Computational time test compiled;</li>
<li> Ongoing: </li>
<ul>
<li> Issue in the distance error (position of the cluster for drivers vs real position);</li>
<li> Compile Rmd for computational clustering quality results.</li>
</ul>
</ul>

<h3>Release of 4/08/2016:</h3>
<ul>
<li>Added function to easily compute Normalized Mutual Information from QuantumClone output - for simulated data;</li>
<li>Added pipelines to compare to paper pipeline.</li>
</ul>
<h3>Release of 29/07/2016:</h3>

<ul>
<li>Option to force single copy status - i.e. integrate QuantumCloneSingle in QuantumClone</li>
<li>Perspectives:</li>
<ul>
<li>tidy plot code in cluster_plot;</li>
<li>Reproducible test in R (QC only);</li>
<li>Reproducible testing in R+bash (QC+sciclone+pyClone)</li>
</ul>  
</ul>  

<h3>Release of 28/07/2016:</h3>
<ul>
<li>Added AIC (Akaike Information Criterion);</li>
<li>Added modified BIC (user input);</li>
<li>Added possibility to use `optimx` instead of `optim` for the maximization step;</li>
<li>Using `compiler` package to improve computing time;</li>
<li>Improved code readability;</li>
<li>Improved parallel computation handling;</li>
<li>Solved conflicting name in simulation: cellularity of clone is now refered to as cellularit;</li>
<li>Added option to keep all iteration models;</li>
</ul>  


<h3>Release of 10/12/2015:</h3>
<ul>  

<li>Changed graph output to pdf;</li>
<li>Fixed non-clustered graph name file</li>
</ul>  

<h3>Release of 12/11/2015:</h3>
<ul>  

<li>Added release tag on github;</li>
<li>Checked for ggplot2 scheduled release</li>
</ul>  

<h3>Release of 2/11/2015:</h3>
<ul>  

<li>Correction of evolution plot;</li>
<li>Adding content to vignettes;</li>
<li>Added phylogenetic tree reconstruction and representation, along with examples;</li>
<li>Added ImportFrom according to notes from R-devel: http://win-builder.r-project.org/;</li>
</ul>  

<h3>Release of 30/10/2015:</h3>

<h4> Major fix </h4>
<ul>  

<li>Post processing affected order of mutations but not of clustering. The clustering thus appeared erroneous in version 0.9.15 and 0.10.15;</li>
</ul>  

<h4> Minor changes </h4>
<ul>  
<li>Adding vignette release;</li>
<li>Changed dependency of doSNOW to doParallel;</li>
<li>Added data to decrease loading and check time;</li>
<li>Added function to plot results from output (by 2 elements);</li>
<li>Added function to plot cellularity of clone in each sample (for time series, with proportion of mutation in each clone);</li>
</ul>  
